ddtrace/sampler.py

```diff
@@ -92,3 +103,266 @@ def set_sample_rate_by_service(self, rate_by_service):
+        self.rules = rules
+
+        # Configure rate limiter
+        if rate_limit is None:
+            rate_limit = self.DEFAULT_RATE_LIMIT
+        self.limiter = RateLimiter(rate_limit)
+        self.default_sampler = SamplingRule(sample_rate=default_sample_rate)
+
+        # TODO: Remove when we no longer use the fallback
+        self._priority_sampler = _priority_sampler
+
+    def _set_priority(self, span, priority):
+        if span._context:
+            span._context.sampling_priority = priority
+        span.sampled = priority is AUTO_KEEP
+
+    def sample(self, span):
+        """
+        Decide whether the provided span should be sampled or not
+
+        The span provided should be the root span in the trace.
+
+        :param span: The root span of a trace
+        :type span: :class:`ddtrace.span.Span`
+        :returns: Whether the span was sampled or not
+        :rtype: :obj:`bool`
+        """
+        # If there are rules defined, then iterate through them and find one that wants to sample
+        matching_rule = None
+        if self.rules:
+            # Go through all rules and grab the first one that matched
+            # DEV: This means rules should be ordered by the user from most specific to least specific
+            for rule in self.rules:
+                if rule.matches(span):
+                    matching_rule = rule
+                    break
+            else:
+                # No rule matches, fallback to priority sampling if set
+                if self._priority_sampler:
+                    if self._priority_sampler.sample(span):
+                        self._set_priority(span, AUTO_KEEP)
+                        return True
+                    else:
+                        self._set_priority(span, AUTO_REJECT)
+                        return False
+
+                # No rule matches, no priority sampler, use the default sampler
+                matching_rule = self.default_sampler
+
+            # Sample with the matching sampling rule
+            if not matching_rule.sample(span):
+                self._set_priority(span, AUTO_REJECT)
+                return False
+            else:
+                # Do not return here, we need to apply rate limit
+                self._set_priority(span, AUTO_KEEP)
+
+        # Ensure all allowed traces adhere to the global rate limit
+        if not self.limiter.is_allowed():
+            self._set_priority(span, AUTO_REJECT)
+            return False
+
+        # We made it by all of checks, sample this trace
+        self._set_priority(span, AUTO_KEEP)
+        return True
+
+
+class SamplingRule(object):
+    """
+    Definition of a sampling rule used by :class:`DatadogSampler` for applying a sample rate on a span
+    """
+    __slots__ = ('_sample_rate', '_sampling_id_threshold', 'service', 'name', 'resource', 'tags')
+
+    NO_RULE = object()
+
+    def __init__(self, sample_rate, service=NO_RULE, name=NO_RULE, resource=NO_RULE, tags=NO_RULE):
```

Comment:
is `NO_RULE` really needed or wouldn't be `None` enough?

Reply:
Unless you want to test for `None`, e.g. `SamplingRule(sample_rate=0.0, service=None)`

Maybe not a super popular case, but consideration.

Reply:
Makes sense.